clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPoolMetricsRegistry.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.producer.internals;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.MetricNameTemplate;
+import org.apache.kafka.common.metrics.Measurable;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Sensor;
+
+public class BufferPoolMetricsRegistry {
```

Comment:
Why do we have so many registry classes instead of a single one for the producer, for example?

Reply:
The primary reason is that some components are used in several places. For example, `org.apache.kafka.common.network.Selector` has a bunch of metrics and is used in the consumer/producer/connect/streams/AdminClient. The `org.apache.kafka.clients.consumer.internals.AbstractCoordinator` is used in both the consumer as well as connect. So I need those registries to be in separate classes so that they can be included in multiple places in the documentation. What that means is, I want to remove the section "[Common monitoring metrics for producer/consumer/connect/streams](http://kafka.apache.org/documentation/#selector_monitoring)", and simply replicate the metrics in each of the producer/consumer/connect/streams sections. I suppose now is a good time to ask, do you think that is okay to get rid of that "common" section?

The other reason was just that I thought it be nice for the registries to parallel the code. So BufferPool metrics in a BufferPoolMetricsRegistry class, etc.

Reply:
I think it's worth thinking in terms of internal versus shared components. For internal components (i.e. only used by one client), I'm not sure how useful it is to separate them. If we do think it's useful to separate them, maybe we should use static inner classes instead.

With regards to the documentation change, it may be worth asking in the mailing list to get more opinions.

Reply:
By "static inner classes," what do you mean? Do you mean that the ProducerMetrics class would have static inner classes, one for each type? Or, do you mean that say the BufferPool.java class would have a static inner class for its metrics?

For the documentation, I'll ask the kafka-users list to get more opinions.